tests/templates/test_subroutines/test_flip_sign.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Tests for the FlipSign template.
+"""
+import pytest
+import numpy as np
+from pennylane import numpy as pnp
+import pennylane as qml
+
+
+class TestFlipSign:
+    """Tests that the template defines the correct sign flip."""
+
+    @pytest.mark.parametrize(
+        ("test_status", "n_qubits"),
+        [
+            ([1, 0, 1, 0], 4),
+            ([1, 0, 1, 0, 1], 5),
+            ([0, 1, 1, 1, 1, 0, 1], 7),
+        ],
+    )
+    def test_eval(self, test_status, n_qubits):
+
+        dev = qml.device("default.qubit", wires=n_qubits, shots=1)
+
+        def circuit_template():
+
+            for wire in list(range(n_qubits)):
+                qml.Hadamard(wires=wire)
+
+            qml.FlipSign(test_status, wires=list(range(n_qubits)))
+
+            for wire in list(range(n_qubits)):
+                qml.Hadamard(wires=wire)
+
+            return qml.sample(qml.PauliZ(n_qubits - 1))
+
```

Comment:
I am not sure if this test is true. The way to check that it works series with a script like this:

```python

import pennylane as qml

dev = qml.device("default.qubit", wires=2)
@qml.qnode(dev)
def circuit():
    for wire in list(range(2)):
        qml.Hadamard(wires = wire)
    qml.FlipSign([1,0], wires = list(range(2)))

    return qml.state()

circuit()
```

```console
tensor([ 0.5+0.j,  0.5+0.j, -0.5+0.j,  0.5+0.j], requires_grad=True)
```

As we can see, the third state (which corresponds to the value [1,0]) is the only one that changes sign :)

Reply:
I have a question here, I cannot find how to assert the output of the circuit, please can you give some help? Thanks!

Reply:
Look this code:

```python
import pennylane as qml
from pennylane import numpy as np

n = 6

dev = qml.device("default.qubit", wires=3)
@qml.qnode(dev)
def circuit():
    for wire in list(range(3)):
        qml.Hadamard(wires = wire)
    qml.FlipSign(n, wires = list(range(3)))

    return qml.state()

# we check that only the indicated value has been changed
for ind, x in enumerate(circuit()):
    if ind == n:
        print(np.sign(x) == -1)
    else:
        print(np.sign(x) == 1)
```

It still needs to be put in test format but if it works well it will print everything `True` s.
On the other hand, I don't think you need it but just in case this script will convert from binary list to integer list.

```python

arr_bin = [1,1,0]

# On this line you can pass the array to binary
n = sum([arr_bin[i]*2**(len(arr_bin)-i-1) for i in range(len(arr_bin))])

```

Reply:
your test does not give me all True on my side, so it cannot pass the test just with state [1,0] and 2 wires can pass the test, something odd happening here👀

Reply:
Definitely something weird is happening, when I set the state in test as (6, 3) works fine but if i set the test as ([1,1,0],3) then fails and the code does the thing properly, I had a look many times but I keep going with the research

Reply:
The test I have sent you only works with integers, if you want to do it in general, in the test itself you would have to check if it is integer or list. In case it is list you apply the second script. But with that the test works with integers is enough because underneath the list processing is executed 😄

Reply:
anyway I had done that before but it did not work as expected, then I just leave the code for the integer, thanks

Reply:
Finally I added the binary array choice and the test passes💪